src/particle_filter/include/particle_filter/distanceTransform.h

```diff
@@ -0,0 +1,177 @@
+#ifndef DISTANCE_TRANSFORM_H
+#define DISTANCE_TRANSFORM_H
+#include <limits>
+
+class distanceTransform
+{
+ public:
+  double voxel_size; //voxel size
+  double world_range[3][2]; //work space range
+  double ***dist_transform;
+  double ***obstacle_map;
+  distanceTransform(double range, double voxel_size);
+  void build(double cube_para[3], double World_Range[3][2]);
+ protected:
+  int voxel_nums[3]; // voxel number along each dimension
+  void distanceTransform_1D(double ***dist_transform_1D, int range[3], double ***cost_fun, int dir, int idx1, int idx2);
+};
+/*distanceTransform::distanceTransform(double World_Range[3][2], double Grid_Res) :voxel_size(Grid_Res) {
+  memcpy(world_range, World_Range, 6 * sizeof(double));
+  }*/
+
+distanceTransform::distanceTransform(double range, double grid_Res): voxel_size(grid_Res) {
+  for (int i = 0; i < 3; i++) {
+    voxel_nums[i] = ceil(range / voxel_size);
+  }
+  dist_transform = new double **[voxel_nums[0]];
+  obstacle_map = new double **[voxel_nums[0]];
+  for (int i = 0; i < voxel_nums[0]; i++) {
+    dist_transform[i] = new double *[voxel_nums[1]];
+    obstacle_map[i] = new double *[voxel_nums[1]];
+    for (int j = 0; j < voxel_nums[1]; j++) {
+      dist_transform[i][j] = new double[voxel_nums[2]];
+      obstacle_map[i][j] = new double[voxel_nums[2]];
+    }
+  }
+}
+
+void distanceTransform::distanceTransform_1D(double ***dist_transform_1D, int range[3], double ***cost_fun, int dir, int idx1, int idx2)
+{
+  int k = 0;
+  int *envelope = new int[range[dir]];
+  envelope[0] = 0;
+  double *bound = new double[range[dir] + 1];
+  bound[0] = -std::numeric_limits<double>::infinity();
+  bound[1] = std::numeric_limits<double>::infinity();
+  double s;
+  if (dir == 0) {
+    for (int q = 1; q < range[dir]; q++) {
```

Comment:
Everything inside this if/else is confusing enough that it should be commented, or rewritten in a simpler way